akka-persistence-typed/src/main/scala/akka/persistence/typed/javadsl/CommandHandler.scala

```diff
@@ -17,159 +18,94 @@ import akka.util.OptionVal
    */
-  def build(): CommandHandler[Command, Event, State] = {
-    val builtCases = cases.reverse.toArray
-    cases = Nil
-    new CommandHandler[Command, Event, State] {
-      override def apply(ctx: ActorContext[Command], state: State, command: Command) = {
-        var idx = 0
-        var effect: OptionVal[Effect[Event, State]] = OptionVal.None
-
-        while (idx < builtCases.length && effect.isEmpty) {
-          val curr = builtCases(idx)
-          if (curr.predicate(command)) {
-            effect = OptionVal.Some(curr.handler.apply(ctx, state, command))
-          }
-          idx += 1
-        }
-
-        effect match {
-          case OptionVal.None    ⇒ throw new MatchError(s"No match found for command of type [${command.getClass.getName}]")
-          case OptionVal.Some(e) ⇒ e.asInstanceOf[EffectImpl[Event, State]]
-        }
-      }
-    }
-  }
+  def builder[Command, Event, State](statePredicate: FI.TypedPredicate[State]): CommandHandlerBuilder[Command, Event, State, State] =
+    new CommandHandlerBuilder(classOf[Any].asInstanceOf[Class[State]], statePredicate)
 
+  /**
+   * INTERNAL API
+   */
+  @InternalApi private final case class CommandHandlerCase[Command, Event, State](
+    commandPredicate: Command ⇒ Boolean,
+    statePredicate:   State ⇒ Boolean,
+    handler:          BiFunction[State, Command, Effect[Event, State]])
 }
-/**
- * Mutable builder for nested Java [[CommandHandler]]s where different states should have different command handlers.
- * [[CommandHandler]] per state are added with the `match` methods and finally a [[CommandHandler]] is created with [[ByStateCommandHandlerBuilder#build]].
- *
- * Match statements are appended and evaluated in order, the first one to match is used. If no match is found when
- * evaluating the built [[CommandHandler]] for a given state a [[scala.MatchError]] is thrown.
- */
-
-final class ByStateCommandHandlerBuilder[Command, Event, State] @InternalApi private[javadsl] () {
 
-  private final case class ByStateCase(predicate: State ⇒ Boolean, handler: CommandHandler[Command, Event, State])
+final class CommandHandlerBuilder[Command, Event, S <: State, State] @InternalApi private[persistence] (
+  val stateClass: Class[S], val statePredicate: FI.TypedPredicate[S]) {
+  import CommandHandlerBuilder.CommandHandlerCase
 
-  private var cases: List[ByStateCase] = Nil
+  private var cases: List[CommandHandlerCase[Command, Event, State]] = Nil
 
-  private def addCase(predicate: State ⇒ Boolean, handler: CommandHandler[Command, Event, State]): Unit = {
-    cases = ByStateCase(predicate, handler) :: cases
+  private def addCase(predicate: Command ⇒ Boolean, handler: BiFunction[S, Command, Effect[Event, State]]): Unit = {
+    cases = CommandHandlerCase[Command, Event, State](
+      commandPredicate = predicate,
+      statePredicate = state ⇒ stateClass.isAssignableFrom(state.getClass) && statePredicate.defined(state.asInstanceOf[S]),
```

Comment:
This isAssignable is useless when we pass in classOf[Any].asInstanceOf but i guess that is okay as it is just the predicate that matters?

Reply:
yes, I think that was my thinking